### ISSUE TITLE
Option for vspipe to output mkv file/stream

### ISSUE THREAD

**Issue by kevina** [vapoursynth/vapoursynth#293]
How difficult would it be to get vspipe to output a primitive Matroska file/stream?  This will be very useful in dealing with variable-frame-rate (or VPR) footage.

I am using vapoursynth to clean up some (raw mjpeg) VFR webcame footage before encoding it for archival purposes.  Right now I am using `ffmpeg` to encode the video, but then I have to use `mkvmerge` to fix the timecodes as ffmpeg does not support using a timecode file. I would like to avoid having to use `mkvmerge`. In addition I would like to be able to fix the keyframe interval to every X seconds (using `-force_key_frames expr:gte(t,n_forced*GOP_LEN_IN_SECONDS)`).  That is not going to work unless `ffmpeg` has the correct timecodes.

Note that is possible to stream a mkv file as ffmpeg can output a mkv stream to stdout by using `ffmpeg ... -f matroska -`.

**Comment by myrsloik** [comment]
That seems like a lot of work just for outputting timestamps...

**Comment by kevina** [comment]
Maybe, but I don't know of any other format the ffmpeg can use that will support VFR video.  I was thinking maybe there is a library out there that could do most of the work for you.

**Comment by myrsloik** [comment]
This is now actually on the roadmap once audio support is in. Once you need to bundle multiple streams it makes a lot more sense.

**Comment by myrsloik** [comment]
Will probably be done using https://cyanreg.github.io/avtransport/ if ffmpeg ever includes it

**Comment by n00mkrad** [comment]
Implemented NUT output which supports audio, VFR, and RGB formats in addition to the YUV formats that y4m can output.

https://github.com/n00mkrad/vapoursynth-mod/releases

**Comment by myrsloik** [comment]
Very interesting. May take me a bit to review.

**Comment by adworacz** [comment]
Bumping this - with all of the new improvements in audio support, including the new issue around splicing/trimming, having proper NUT muxing would be very helpful. 

I will note, there is a workaround when using something like FFMPEG. You can use Python threads and FFMPEG’s multiple inputs with pipes to process everything at once. I have a working solution, just not at that computer right now. 

Still, being able to mux to NUT (especially via the Python `output()` API) would be a great convenience.

**Comment by myrsloik** [comment]
Right, I should deal with this one next. Maybe I'll simply turn it into a separate command. Depends on how tangled the code becomes.

**Comment by n00mkrad** [comment]
I also added multi-stream output, so you can do fancy stuff like writing to multiple output files at once or merging alpha without intermediate files.

**Comment by myrsloik** [comment]
I thought the nut implementation looked a bit complicated so with the help of AI I cooked up this mkv branch that does the same in like half the lines of code:
https://github.com/vapoursynth/vapoursynth/tree/vspipemkv

I still have some testing to do but it looks very promising so far

**Comment by n00mkrad** [comment]
@myrsloik Appreciate it. I think the VFR/timestamp stuff got pretty messy. I don't claim that mine is better in any way; I just had to resort to LLMs because I have near-zero cpp knowledge.

**Comment by myrsloik** [comment]
Merged now. Go test the normal master artifacts
